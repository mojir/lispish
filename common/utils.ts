import { FunctionType } from '../src/constants/constants'
import type { UnknownRecord } from '../src/interface'
import { isLitsFunction } from '../src/typeGuards/litsFunction'

export function throttle<T extends (...args: any[]) => void>(func: T) {
  let openForBusiness = true
  return function (this: any, ...args: Parameters<T>) {
    if (openForBusiness) {
      requestAnimationFrame(() => openForBusiness = true)
      openForBusiness = false
      func.apply(this, args)
    }
  }
}

export function stringifyValue(value: unknown, html: boolean): string {
  const gt = html ? '&gt;' : '>'
  const lt = html ? '&lt;' : '<'
  if (isLitsFunction(value)) {
    if (value.t === FunctionType.Builtin)
      return `${lt}builtin function ${value.n}${gt}`
    else
      return `${lt}function ${(value as unknown as UnknownRecord).n ?? '\u03BB'}${gt}`
  }
  if (value === null)
    return 'null'

  if (typeof value === 'object' && value instanceof Error)
    return value.toString()

  if (typeof value === 'object' && value instanceof RegExp)
    return `${value}`

  if (value === Number.POSITIVE_INFINITY)
    return `${Number.POSITIVE_INFINITY}`

  if (value === Number.NEGATIVE_INFINITY)
    return `${Number.NEGATIVE_INFINITY}`

  if (typeof value === 'number' && Number.isNaN(value))
    return 'NaN'

  return JSON.stringify(value, null, 2)
}

export function findAllOccurrences(input: string, pattern: RegExp): Set<string> {
  const matches = [...input.matchAll(pattern)]
  return new Set(matches.map(match => match[0]))
}
