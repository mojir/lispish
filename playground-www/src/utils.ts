export function isNotNull<T>(value: T | null | undefined): value is T {
  return value !== null && value !== undefined
}

export function assertNotNull<T>(value: T | null | undefined): asserts value is T {
  if (!isNotNull(value))
    throw new Error('Value is null or undefined')
}

export function asNotNull<T>(value: T | null | undefined): T {
  assertNotNull(value)
  return value
}
