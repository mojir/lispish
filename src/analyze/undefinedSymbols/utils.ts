import { UndefinedSymbolEntry } from './interface'

export function joinUndefinedSymbols(
  ...undefinedSymbolsSets: Array<Set<UndefinedSymbolEntry>>
): Set<UndefinedSymbolEntry> {
  const undefinedSymbols = new Set<UndefinedSymbolEntry>()

  for (const set of undefinedSymbolsSets) {
    set.forEach(symbol => undefinedSymbols.add(symbol))
  }
  return undefinedSymbols
}

export function addAnalyzeResults(target: Set<UndefinedSymbolEntry>, source: Set<UndefinedSymbolEntry>): void {
  source.forEach(symbol => target.add(symbol))
}
