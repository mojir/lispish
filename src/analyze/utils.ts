import type { AnalyzeResult } from './interface'

export function joinAnalyzeResults(...results: AnalyzeResult[]): AnalyzeResult {
  const result: AnalyzeResult = {
    undefinedSymbols: new Set(),
  }
  for (const input of results)
    input.undefinedSymbols.forEach(symbol => result.undefinedSymbols.add(symbol))

  return result
}

export function addAnalyzeResults(target: AnalyzeResult, source: AnalyzeResult): void {
  source.undefinedSymbols.forEach(symbol => target.undefinedSymbols.add(symbol))
}
