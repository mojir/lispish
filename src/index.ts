export { AstNodeType, TokenType, FunctionType } from './constants/constants'
export {
  isBuiltinFunction,
  isLitsFunction,
  asLitsFunction,
  assertLitsFunction,
  isUserDefinedFunction,
  asUserDefinedFunction,
  assertUserDefinedFunction,
  isNativeJsFunction,
  asNativeJsFunction,
  assertNativeJsFunction,
} from './typeGuards/litsFunction'
export { type LitsFunction, type NativeJsFunction, type ExtraData } from './parser/interface'
export type { Context } from './evaluator/interface'
export type { Ast } from './parser/interface'
export type { Token, SourceCodeInfo } from './tokenizer/interface'
export { normalExpressionKeys, specialExpressionKeys } from './builtin'
export { reservedNames } from './reservedNames'
export { Lits } from './Lits/Lits'
export { type LitsError, isLitsError } from './errors'
export type { LitsParams, LitsRuntimeInfo, LazyValue } from './Lits/Lits'
export { createNativeJsFunction } from './utils'
