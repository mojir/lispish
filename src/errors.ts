import { Arr } from './interface'
import { DebugInfo } from './tokenizer/interface'
import { getCodeMarker, valueToString } from './utils/helpers'

function getLitsErrorMessage(message: string, debugInfo?: DebugInfo) {
  return `${message}${
    debugInfo ? `\n${debugInfo === `EOF` ? `EOF` : `${debugInfo.code}\n${getCodeMarker(debugInfo)}`}` : ``
  }`
}

export class RecurSignal extends Error {
  public params: Arr
  constructor(params: Arr) {
    super(`recur, params: ${params}`)
    Object.setPrototypeOf(this, RecurSignal.prototype)
    this.name = `RecurSignal`
    this.params = params
  }
}

export abstract class AbstractLitsError extends Error {
  public debugInfo?: DebugInfo
  public shortMessage: string
  constructor(message: string | Error, debugInfo?: DebugInfo) {
    if (message instanceof Error) {
      message = `${message.name}${message.message ? `: ${message.message}` : ``}`
    }
    super(getLitsErrorMessage(message, debugInfo))
    this.shortMessage = message
    this.debugInfo = debugInfo
    Object.setPrototypeOf(this, AbstractLitsError.prototype)
    this.name = `AbstractLitsError`
  }
}

export class LitsError extends AbstractLitsError {
  constructor(message: string | Error, debugInfo?: DebugInfo) {
    super(message, debugInfo)
    Object.setPrototypeOf(this, LitsError.prototype)
    this.name = `LitsError`
  }
}

export class NotAFunctionError extends AbstractLitsError {
  constructor(fn: unknown, debugInfo?: DebugInfo) {
    const message = `Expected function, got ${valueToString(fn)}.`
    super(message, debugInfo)
    Object.setPrototypeOf(this, NotAFunctionError.prototype)
    this.name = `NotAFunctionError`
  }
}

export class UserDefinedError extends AbstractLitsError {
  constructor(message: string | Error, debugInfo?: DebugInfo) {
    super(message, debugInfo)
    Object.setPrototypeOf(this, UserDefinedError.prototype)
    this.name = `UserDefinedError`
  }
}

export class AssertionError extends AbstractLitsError {
  constructor(message: string | Error, debugInfo?: DebugInfo) {
    super(message, debugInfo)
    Object.setPrototypeOf(this, AssertionError.prototype)
    this.name = `AssertionError`
  }
}

export class UndefinedSymbolError extends AbstractLitsError {
  public symbol: string
  constructor(symbolName: string, debugInfo?: DebugInfo) {
    const message = `Undefined symbol '${symbolName}'.`
    super(message, debugInfo)
    this.symbol = symbolName
    Object.setPrototypeOf(this, UndefinedSymbolError.prototype)
    this.name = `UndefinedSymbolError`
  }
}
