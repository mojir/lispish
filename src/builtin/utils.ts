import { LitsError } from '../errors'
import type { ContextStack } from '../evaluator/ContextStack'
import type { AstNode, BindingNode } from '../parser/interface'
import { postfixReservedNamesRecord } from '../tokenizer/postfix/postfixReservedNames'
import type { SourceCodeInfo } from '../tokenizer/interface'
import type { Builtin } from './interface'
import type { SpecialExpressionName } from '.'

export type Arity = number | { min: number }

export interface FunctionOverload {
  as: FunctionArguments
  a: Arity
  b: AstNode[]
}

export interface FunctionArguments {
  m: string[]
  r?: string
  b: BindingNode[]
}

export function assertNameNotDefined<T>(
  name: T,
  contextStack: ContextStack,
  builtin: Builtin,
  sourceCodeInfo?: SourceCodeInfo,
): asserts name is T {
  if (typeof name !== 'string')
    return

  if (builtin.specialExpressions[name as SpecialExpressionName])
    throw new LitsError(`Cannot define variable ${name}, it's a special expression.`, sourceCodeInfo)

  if (builtin.normalExpressions[name])
    throw new LitsError(`Cannot define variable ${name}, it's a builtin function.`, sourceCodeInfo)

  // eslint-disable-next-line ts/no-unsafe-member-access
  if ((postfixReservedNamesRecord as any)[name])
    throw new LitsError(`Cannot define variable ${name}, it's a reserved name.`, sourceCodeInfo)

  if (contextStack.globalContext[name])
    throw new LitsError(`Name already defined "${name}".`, sourceCodeInfo)
}
