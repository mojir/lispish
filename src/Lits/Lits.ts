import { analyzeAst } from '../analyze'
import type { AnalyzeResult } from '../analyze/interface'
import { builtin, normalExpressionKeys, specialExpressionKeys } from '../builtin'
import { FunctionType } from '../constants/constants'
import { evaluate } from '../evaluator'
import { ContextStack } from '../evaluator/ContextStack'
import type { Context } from '../evaluator/interface'
import type { Any, Obj } from '../interface'
import { parse } from '../parser'
import type { Ast, LitsFunction, NativeJsFunction } from '../parser/interface'
import { tokenize } from '../tokenizer'
import type { TokenStream } from '../tokenizer/interface'
import { Cache } from './Cache'

export interface LitsRuntimeInfo {
  astCache: Cache | null
  astCacheSize: number | null
  debug: boolean
}

export interface LazyValue {
  read: () => unknown
  [key: string]: unknown
}

export interface JsFunction {
  fn: (...args: any[]) => unknown
  [key: string]: unknown
}

export interface LitsParams {
  globalContext?: Context
  contexts?: Context[]
  values?: Record<string, unknown>
  lazyValues?: Record<string, LazyValue>
  jsFunctions?: Record<string, JsFunction>
  filePath?: string
}

interface LitsConfig {
  initialCache?: Record<string, Ast>
  astCacheSize?: number | null
  debug?: boolean
}

export class Lits {
  private astCache: Cache | null
  private astCacheSize: number | null
  private debug: boolean

  constructor(config: LitsConfig = {}) {
    this.debug = config.debug ?? false
    this.astCacheSize = config.astCacheSize ?? null
    if (this.astCacheSize) {
      this.astCache = new Cache(this.astCacheSize)
      const initialCache = config.initialCache ?? {}
      for (const cacheEntry of Object.keys(initialCache))
        this.astCache.set(cacheEntry, initialCache[cacheEntry] as Ast)
    }
    else {
      this.astCache = null
    }
  }

  public getRuntimeInfo(): LitsRuntimeInfo {
    return {
      astCacheSize: this.astCacheSize,
      astCache: this.astCache,
      debug: this.debug,
    }
  }

  public run(program: string, params: LitsParams = {}): unknown {
    const ast = this.generateAst(program, params.filePath)
    const result = this.evaluate(ast, params)
    return result
  }

  public context(program: string, params: LitsParams = {}): Context {
    const contextStack = createContextStack(params)
    const ast = this.generateAst(program, params.filePath)
    evaluate(ast, contextStack)
    return contextStack.globalContext
  }

  public analyze(program: string): AnalyzeResult {
    const params: LitsParams = {}
    const contextStack = createContextStack(params)
    const ast = this.generateAst(program, params.filePath)

    return analyzeAst(ast.b, contextStack, builtin)
  }

  public tokenize(program: string, filePath?: string): TokenStream {
    return tokenize(program, { debug: this.debug, filePath })
  }

  public parse(tokenStream: TokenStream): Ast {
    return parse(tokenStream)
  }

  private evaluate(ast: Ast, params: LitsParams): Any {
    const contextStack = createContextStack(params)
    return evaluate(ast, contextStack)
  }

  public apply(fn: LitsFunction, fnParams: unknown[], params: LitsParams = {}): Any {
    const fnName = 'FN_2eb7b316-471c-5bfa-90cb-d3dfd9164a59'
    const paramsString: string = fnParams
      .map((_, index) => {
        return `${fnName}_${index}`
      })
      .join(' ')
    const program = `(${fnName} ${paramsString})`
    const ast = this.generateAst(program, params.filePath)

    const hostValues: Obj = fnParams.reduce(
      (result: Obj, param, index) => {
        result[`${fnName}_${index}`] = param
        return result
      },
      { [fnName]: fn },
    )

    params.values = { ...params.values, ...hostValues }

    return this.evaluate(ast, params)
  }

  private generateAst(untrimmedProgram: string, filePath?: string): Ast {
    const program = untrimmedProgram.trim()

    if (this.astCache) {
      const cachedAst = this.astCache.get(program)
      if (cachedAst)
        return cachedAst
    }
    const tokenStream = this.tokenize(program, filePath)
    const ast: Ast = this.parse(tokenStream)
    this.astCache?.set(program, ast)
    return ast
  }
}

export function createContextStack(params: LitsParams = {}): ContextStack {
  const globalContext = params.globalContext ?? {}
  // Contexts are checked from left to right
  const contexts = params.contexts ? [globalContext, ...params.contexts] : [globalContext]
  const contextStack = new ContextStack({
    contexts,
    values: params.values,
    lazyValues: params.lazyValues,
    nativeJsFunctions:
      params.jsFunctions
      && Object.entries(params.jsFunctions).reduce((acc: Record<string, NativeJsFunction>, [name, jsFunction]) => {
        if (specialExpressionKeys.includes(name)) {
          console.warn(`Cannot shadow special expression "${name}", ignoring.`)
          return acc
        }
        if (normalExpressionKeys.includes(name)) {
          console.warn(`Cannot shadow builtin function "${name}", ignoring.`)
          return acc
        }
        acc[name] = {
          t: FunctionType.NativeJsFunction,
          f: jsFunction,
          n: name,
          __fn: true,
        }
        return acc
      }, {}),
  })
  return contextStack
}
