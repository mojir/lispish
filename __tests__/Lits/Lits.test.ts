import { Ast, Lits } from '../../src'
import { UndefinedSymbolError } from '../../src/errors'
import { Cache } from '../../src/Lits/Cache'
import { litsFunction } from '../../src/utils/assertion'

describe(`TEST`, () => {
  let lits: Lits
  beforeEach(() => {
    lits = new Lits({ debug: true, astCacheSize: 0 })
  })
  test(`without params`, () => {
    const fn = lits.run(`#(+ %1 %2)`)
    litsFunction.assert(fn)
    expect(lits.apply(fn, [2, 3])).toBe(5)
  })
  test(`with empty params`, () => {
    const fn = lits.run(`#(+ %1 %2)`)
    litsFunction.assert(fn)
    expect(lits.apply(fn, [2, 3], {})).toBe(5)
  })

  test(`with params`, () => {
    const fn = lits.run(`#(+ %1 %2 x)`)
    litsFunction.assert(fn)
    expect(lits.apply(fn, [2, 3], { globalContext: { x: { value: 1 } } })).toBe(6)
  })
})

describe(`runtime info`, () => {
  test(`getRuntimeInfo().`, () => {
    const lits = new Lits()
    expect(lits.getRuntimeInfo()).toMatchSnapshot()
  })
  test(`getRuntimeInfo() with ast cache > 0`, () => {
    const lits = new Lits({ astCacheSize: 10 })
    expect(lits.getRuntimeInfo()).toMatchSnapshot()
  })
  test(`getRuntimeInfo() with ast cache = 0`, () => {
    const lits = new Lits({ astCacheSize: 0 })
    expect(lits.getRuntimeInfo()).toMatchSnapshot()
  })
})

describe(`context`, () => {
  let lits: Lits
  beforeEach(() => {
    lits = new Lits({ debug: true })
  })
  test(`a function.`, () => {
    lits = new Lits({ astCacheSize: 10 })
    const contexts = [lits.context(`(defn tripple [x] (* x 3))`)]
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
  })

  test(`a function - no cache`, () => {
    lits = new Lits({ debug: true })
    const contexts = [lits.context(`(defn tripple [x] (* x 3))`, {})]
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
    expect(lits.run(`(tripple 10)`, { contexts })).toBe(30)
  })

  test(`a function - initial cache`, () => {
    const initialCache: Record<string, Ast> = {
      '(pow 2 4)': {
        type: `Program`,
        body: [
          {
            type: `NormalExpression`,
            name: `pow`,
            params: [
              {
                type: `Number`,
                value: 2,
                token: {
                  type: `number`,
                  value: `2`,
                },
              },
              {
                type: `Number`,
                value: 4,
                token: {
                  type: `number`,
                  value: `4`,
                },
              },
            ],
            token: {
              type: `name`,
              value: `pow`,
            },
          },
        ],
      },
    }
    lits = new Lits({ astCacheSize: 10, initialCache })
    expect(lits.run(`(pow 2 2)`)).toBe(4)
    expect(lits.run(`(pow 2 4)`)).toBe(16)
  })

  test(`a variable.`, () => {
    const contexts = [lits.context(`(def magicNumber 42)`)]
    expect(lits.run(`magicNumber`, { contexts })).toBe(42)
  })

  test(`a variable - again.`, () => {
    const contexts = [
      lits.context(`
    (defn zip? [string] (boolean (match (regexp "^\\d{5}$") string)))
    (defn isoDateString? [string] (boolean (match (regexp "^\\d{4}-\\d{2}-\\d{2}$") string)))
    (def NAME_LENGTH 100)
    `),
    ]
    expect(lits.run(`NAME_LENGTH`, { contexts })).toBe(100)
  })

  test(`change imported variable`, () => {
    const contexts = [lits.context(`(def magicNumber 42)`)]
    expect(lits.run(`magicNumber`, { contexts })).toBe(42)
  })

  test(`a function with a built in normal expression name`, () => {
    expect(() => lits.context(`(defn inc (x) (+ x 1))`)).toThrow()
    expect(() => lits.context(`(defn inc (x) (+ x 1))`, { globalContext: {} })).toThrow()
    expect(() => lits.context(`(defn inc (x) (+ x 1))`, { globals: {} })).toThrow()
  })

  test(`a function with a built in special expression name`, () => {
    expect(() => lits.context(`(defn and (x y) (* x y))`)).toThrow()
  })

  test(`a variable twice`, () => {
    const contexts = [lits.context(`(def magicNumber 42) (defn getMagic [] 42)`)]
    expect(() => lits.context(`(def magicNumber 42) (defn getMagic [] 42)`, { contexts })).not.toThrow()
  })

  test(`more than one`, () => {
    const contexts = [lits.context(`(defn tripple [x] (* x 3))`), lits.context(`(def magicNumber 42)`)]
    expect(lits.run(`(tripple magicNumber)`, { contexts })).toBe(126)
  })
})

function ast(n: number): Ast {
  return {
    body: [
      {
        type: `Number`,
        value: n,
      },
    ],
    type: `Program`,
  }
}

describe(`Cache`, () => {
  test(`cannot set same key twice`, () => {
    const cache = new Cache(10)
    cache.set(`a`, ast(1))
    expect(() => cache.set(`a`, ast(2))).toThrow()
  })

  test(`getContent`, () => {
    const cache = new Cache(10)
    cache.set(`a`, ast(1))
    cache.set(`b`, ast(2))
    expect(cache.getContent()).toEqual({
      a: ast(1),
      b: ast(2),
    })
  })

  test(`max cache size must be at least 1`, () => {
    expect(() => new Cache(-1)).toThrow()
    expect(() => new Cache(0)).toThrow()
    expect(() => new Cache(0.1)).not.toThrow()
    expect(() => new Cache(1)).not.toThrow()
  })

  test(`Add an entry.`, () => {
    const cache = new Cache(10)
    expect(cache.size).toBe(0)
    cache.set(`a`, ast(1))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toEqual(ast(1))
    expect(cache.has(`a`)).toBe(true)
  })

  test(`Clear cache.`, () => {
    const cache = new Cache(10)
    cache.set(`a`, ast(1))
    cache.set(`b`, ast(2))
    cache.set(`c`, ast(3))
    expect(cache.size).toBe(3)
    cache.clear()
    expect(cache.size).toBe(0)
  })

  test(`Add an entry - cacheSize = 1`, () => {
    const cache = new Cache(1)
    expect(cache.size).toBe(0)
    cache.set(`a`, ast(1))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toEqual(ast(1))
  })
  test(`maxSize.`, () => {
    const cache = new Cache(1)
    cache.set(`a`, ast(1))
    expect(cache.get(`a`)).toEqual(ast(1))
    cache.set(`b`, ast(2))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toBeUndefined()
    expect(cache.has(`a`)).toBe(false)
    expect(cache.get(`b`)).toEqual(ast(2))
    expect(cache.has(`b`)).toBe(true)
  })
  test(`maxSize.`, () => {
    const cache = new Cache(1)
    cache.set(`a`, ast(1))
    expect(cache.get(`a`)).toEqual(ast(1))
    cache.set(`b`, ast(2))
    expect(cache.size).toBe(1)
    expect(cache.get(`a`)).toBeUndefined()
    expect(cache.has(`a`)).toBe(false)
    expect(cache.get(`b`)).toEqual(ast(2))
    expect(cache.has(`b`)).toBe(true)
  })
})

describe(`regressions`, () => {
  let lits: Lits
  beforeEach(() => {
    lits = new Lits({ debug: true })
  })
  test(`debugInfo`, () => {
    try {
      lits.run(`(loop [n 3]
  (write! n)
  (when (not zero? n))
    (recur (dec n))
  )
)`)
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.line).toBe(3)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.column).toBe(10)
    }
  })
  test(`name not recognized`, () => {
    expect(() => lits.run(`(asd)`)).toThrowError(UndefinedSymbolError)
    expect(() => lits.run(`asd`)).toThrowError(UndefinedSymbolError)
  })

  test(`debug info when executing function with error in`, () => {
    const program = `(defn formatPhoneNumber [$data]
  (if (string? $data)
    (let [phoneNumber (if (= "+" (nth $data 0)) (subs $data 2) $data)]
      (cond
        (> (count phoneNumber) 6)
          (astr
            "("
            (subs phoneNumber 0 3)

            ") "
            (subs phoneNumber 3 6)
            "-"
            (subs phoneNumber 6))

        (> (count phoneNumber) 3)
          (str "(" (subs phoneNumber 0 3) ") " (subs phoneNumber 3))

        (> (count phoneNumber) 0)
          (str "(" (subs phoneNumber 0))

        true
          phoneNumber
      )
    )
    ""
  )
)

(formatPhoneNumber "+1234232123")`
    try {
      lits.run(program)
      fail()
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.line).toBe(6)
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      expect((error as any).debugInfo.column).toBe(12)
    }
  })
  test(`unexpected argument`, () => {
    try {
      lits.run(`(+ 1 + 2)`)
    } catch (error) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const anyError = error as any
      expect(anyError.debugInfo.line).toBe(1)
      expect(anyError.debugInfo.column).toBe(6)
    }
  })
})
