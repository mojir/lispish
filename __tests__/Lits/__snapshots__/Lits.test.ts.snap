// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runtime info > getRuntimeInfo() with ast cache = 0 1`] = `
{
  "astCache": null,
  "astCacheSize": 0,
  "debug": false,
}
`;

exports[`runtime info > getRuntimeInfo() with ast cache > 0 1`] = `
{
  "astCache": Cache {
    "_size": 0,
    "cache": {},
    "firstEntry": undefined,
    "lastEntry": undefined,
    "maxSize": 10,
  },
  "astCacheSize": 10,
  "debug": false,
}
`;

exports[`runtime info > getRuntimeInfo(). 1`] = `
{
  "astCache": null,
  "astCacheSize": null,
  "debug": false,
}
`;
